src/main/telemetry/esc_telemetry.c

```diff
@@ -0,0 +1,330 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <platform.h>
+
+#include "fc/config.h"
+#include "config/feature.h"
+#include "config/config_master.h"
+
+#include "common/utils.h"
+
+#include "drivers/system.h"
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+#include "drivers/pwm_output.h"
+#include "io/serial.h"
+
+#include "flight/mixer.h"
+
+#include "sensors/battery.h"
+
+#include "esc_telemetry.h"
+
+#include "build/debug.h"
+
+/*
+KISS ESC TELEMETRY PROTOCOL
+---------------------------
+
+One transmission will have 10 times 8-bit bytes sent with 115200 baud and 3.6V.
+
+Byte 0: Temperature
+Byte 1: Voltage high byte
+Byte 2: Voltage low byte
+Byte 3: Current high byte
+Byte 4: Current low byte
+Byte 5: Consumption high byte
+Byte 6: Consumption low byte
+Byte 7: Rpm high byte
+Byte 8: Rpm low byte
+Byte 9: 8-bit CRC
+
+*/
+
+/*
+DEBUG INFORMATION
+-----------------
+
+set debug_mode = DEBUG_ESC_TELEMETRY in cli
+
+0: current motor index requested
+1: number of timeouts
+2: voltage
+3: current
+*/
+
+typedef struct {
+    bool skipped;
+    uint8_t temperature;
+    uint16_t voltage;
+    uint16_t current;
+    uint16_t consumption;
+    uint16_t rpm;
+} esc_telemetry_t;
+
+typedef enum {
+    ESC_TLM_FRAME_PENDING = 1 << 0,     // 1
+    ESC_TLM_FRAME_COMPLETE = 1 << 1     // 2
+} escTlmFrameState_t;
+
+typedef enum {
+    ESC_TLM_TRIGGER_WAIT = 0,
+    ESC_TLM_TRIGGER_READY = 1 << 0,     // 1
+    ESC_TLM_TRIGGER_PENDING = 1 << 1,   // 2
+} escTlmTriggerState_t;
+
+#define ESC_TLM_BAUDRATE 115200
+#define ESC_TLM_BUFFSIZE 10
+#define ESC_BOOTTIME 3000               // 3 seconds
+#define ESC_REQUEST_TIMEOUT 1000        // 1 seconds
+
+static bool tlmFrameDone = false;
+static bool firstCycleComplete = false;
+static uint8_t tlm[ESC_TLM_BUFFSIZE] = { 0, };
+static uint8_t tlmFramePosition = 0;
+static serialPort_t *escTelemetryPort = NULL;
+static esc_telemetry_t escTelemetryData[4];
+static uint32_t escTriggerTimestamp = -1;
+
+static uint8_t escTelemetryMotor = 99;      // motor index 0 - 3
+static bool escTelemetryEnabled = false;
+static escTlmTriggerState_t escTelemetryTriggerState = ESC_TLM_TRIGGER_WAIT;
+
+static uint16_t escVbat = 0;
+static uint16_t escCurrent = 0;
+static uint16_t escConsumption = 0;
+
+static void escTelemetryDataReceive(uint16_t c);
+static uint8_t update_crc8(uint8_t crc, uint8_t crc_seed);
+static uint8_t get_crc8(uint8_t *Buf, uint8_t BufLen);
+static void selectNextMotor(void);
+
+static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+
+bool isEscTelemetryActive(void)
+{
+    return escTelemetryEnabled;
+}
+
+uint16_t getEscTelemetryVbat(void)
+{
+    return escVbat / 10;
+}
+
+uint16_t getEscTelemetryCurrent(void)
+{
+    return escCurrent;
+}
+
+uint16_t getEscTelemetryConsumption(void)
+{
+    return escConsumption;
+}
+
+void pwmRequestTelemetry(uint8_t index)
```

Comment:
if this method kept a track of which motor it had requested telemetry for , then it could simply be called from the telemetry invocation from scheduler - exiting directly if there is still a wait for telemetry from a previous request (until a timeout occurs)